File: packages/ios/Sources/CivUI/CivCurrency.swift
// CivUI — CivCurrency for SwiftUI
// Accessible currency input with dollar prefix and decimal keyboard.
// Delegates to CivTextInput with currency mask configuration.

import SwiftUI

/// Accessible currency input for government applications.
///
/// Wraps `CivTextInput` with currency mask, decimal keyboard, and
/// dollar sign prefix. Mirrors the web `civ-text-input[mask="currency"]` pattern.
///
/// Usage:
/// ```swift
/// CivCurrency(value: $amount)
/// CivCurrency(label: "Monthly income", value: $income, isRequired: true)
/// ```
public struct CivCurrency: View {
    // MARK: - Properties

    /// Visible label text (defaults to "Amount").
    public let label: String

    /// Bound input value (stores raw numeric string, e.g. "1234.56").
    @Binding public var value: String

    /// Help text shown below the label.
    public var hint: String?

    /// Error message. When set, renders with VoiceOver announcement.
    public var error: String?

    /// Whether the field is required.
    public var isRequired: Bool

    /// Whether the field is disabled.
    public var isDisabled: Bool

    /// Whether the field is read-only.
    public var isReadonly: Bool

    /// Called on committed change / focus loss (parallels `civ-change` event).
    public var onChange: ((String) -> Void)?

    /// Called for analytics tracking (parallels `civ-analytics` event).
    public var onAnalytics: ((String, [String: Any]?) -> Void)?

    /// Optional form state for centralized validation.
    public var formState: CivFormState?

    /// Field name for form state registration.
    public var formName: String?

    /// Custom required message for form validation.
    public var requiredMessage: String?

    /// Custom validation function. Returns error string or nil.
    public var formValidate: (() -> String?)?

    /// Whether the field contains PII.
    public var isPii: Bool

    // MARK: - Initializer

    public init(
        label: String = "Amount",
        value: Binding<String>,
        hint: String? = nil,
        error: String? = nil,
        isRequired: Bool = false,
        isDisabled: Bool = false,
        isReadonly: Bool = false,
        onChange: ((String) -> Void)? = nil,
        onAnalytics: ((String, [String: Any]?) -> Void)? = nil,
        formState: CivFormState? = nil,
        formName: String? = nil,
        requiredMessage: String? = nil,
        formValidate: (() -> String?)? = nil,
        isPii: Bool = false
    ) {
        self.label = label
        self._value = value
        self.hint = hint
        self.error = error
        self.isRequired = isRequired
        self.isDisabled = isDisabled
        self.isReadonly = isReadonly
        self.onChange = onChange
        self.onAnalytics = onAnalytics
        self.formState = formState
        self.formName = formName
        self.requiredMessage = requiredMessage
        self.formValidate = formValidate
        self.isPii = isPii
    }

    // MARK: - Body

    public var body: some View {
        CivTextInput(
            label: label,
            value: $value,
            hint: hint,
            error: error,
            isRequired: isRequired,
            isDisabled: isDisabled,
            isReadonly: isReadonly,
            mask: CivInputMask.currency,
            onChange: onChange,
            onAnalytics: onAnalytics,
            formState: formState,
            formName: formName,
            requiredMessage: requiredMessage,
            formValidate: formValidate,
            isPii: isPii
        )
    }
}

// MARK: - Preview

#if DEBUG
struct CivCurrency_Previews: PreviewProvider {
    struct PreviewWrapper: View {
        @State private var amount = ""
        @State private var income = "4500.00"

        var body: some View {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    CivCurrency(
                        value: $amount,
                        isRequired: true
                    )

                    CivCurrency(
                        label: "Monthly income",
                        value: $income,
                        hint: "Before taxes"
                    )

                    CivCurrency(
                        value: .constant(""),
                        error: "Enter an amount",
                        isRequired: true
                    )

                    CivCurrency(
                        label: "Approved amount",
                        value: .constant("12500.00"),
                        isReadonly: true
                    )

                    CivCurrency(
                        value: .constant(""),
                        isDisabled: true
                    )
                }
                .padding()
            }
        }
    }

    static var previews: some View {
        PreviewWrapper()
            .previewDisplayName("Light")
        PreviewWrapper()
            .preferredColorScheme(.dark)
            .previewDisplayName("Dark")
    }
}
#endif

File: packages/android/src/main/kotlin/gov/civui/components/CivCurrency.kt
// CivUI — CivCurrency for Jetpack Compose
// Accessible currency/dollar amount input with formatting.
// Thin wrapper around CivTextInput with currency-specific configuration.

package gov.civui.components

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp

/**
 * Accessible currency input for government applications.
 *
 * Wraps [CivTextInput] with currency mask (dollar prefix, decimal formatting).
 *
 * Usage:
 * ```kotlin
 * var amount by remember { mutableStateOf("") }
 * CivCurrency(
 *     value = amount,
 *     onValueChange = { amount = it },
 * )
 * ```
 */
@Composable
fun CivCurrency(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    label: String = "Amount",
    hint: String? = null,
    error: String? = null,
    required: Boolean = false,
    disabled: Boolean = false,
    readonly: Boolean = false,
    onChange: ((String) -> Unit)? = null,
    onAnalytics: ((event: String, data: Map<String, Any>?) -> Unit)? = null,
    name: String = "",
    formState: CivFormState? = null,
    decimals: Int = 2,
    min: Double? = null,
    max: Double? = null,
    allowNegative: Boolean = false,
) {
    CivTextInput(
        label = label,
        value = value,
        onValueChange = onValueChange,
        modifier = modifier,
        hint = hint,
        error = error,
        required = required,
        disabled = disabled,
        readonly = readonly,
        mask = CivInputMask.Currency,
        onChange = onChange,
        onAnalytics = onAnalytics,
        name = name,
        formState = formState,
    )
}

// MARK: - Preview

@Preview(showBackground = true, name = "CivCurrency")
@Composable
private fun CivCurrencyPreview() {
    Column(modifier = Modifier.padding(16.dp)) {
        var amount by remember { mutableStateOf("") }
        CivCurrency(
            value = amount,
            onValueChange = { amount = it },
            label = "Monthly income",
            required = true,
        )

        CivCurrency(
            value = "1500.00",
            onValueChange = {},
            label = "Estimated cost",
            hint = "Enter the total amount in US dollars",
        )

        CivCurrency(
            value = "",
            onValueChange = {},
            label = "Amount owed",
            error = "Please enter a valid dollar amount",
        )

        CivCurrency(
            value = "",
            onValueChange = {},
            disabled = true,
        )
    }
}
